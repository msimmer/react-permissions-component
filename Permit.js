import React from 'react'

class Permit extends React.Component {
  getPermissions = () => {
    const { permissions } = this.props
    const [u, g, o] = String(permissions)
      .split('')
      .map(a =>
        Number(a)
          .toString(2)
          .padStart('3', '0'),
      )
    return [u, g, o]
  }

  getRole = () => {
    const { role } = this.props
    return this.props.roles[role]
  }

  getAccess = () => {
    const permissions = this.getPermissions()
    const role = this.getRole()
    const [r, w, x] = permissions[role].split('').map(a => Number(a))
    return { r, w, x }
  }

  viewable = () => this.getAccess().r

  editable = () => this.getAccess().w

  executable = () => this.getAccess().x

  children = () =>
    this.editable() || this.viewable()
      ? React.cloneElement(this.props.children)
      : null

  content = ({ editable, viewable, executable }) => {
    return this.props.children
      ? this.children()
      : this.executable() && executable
      ? React.cloneElement(executable)
      : this.editable() && editable
      ? React.cloneElement(editable)
      : this.viewable() && viewable
      ? React.cloneElement(viewable)
      : null
  }

  render = () => <React.Fragment>{this.content(this.props)}</React.Fragment>
}

export default Permit
