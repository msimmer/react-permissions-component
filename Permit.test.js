import React from 'react'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import { configure } from 'enzyme'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
// import { expect } from 'chai'
import Permit from './Permit'

configure({ adapter: new Adapter() })

const Viewable = () => (
  <form>
    <input type="text" readOnly={true} />
  </form>
)
const Editable = () => (
  <form>
    <input type="text" />
    <button value="save" />
  </form>
)
const Executable = () => (
  <form>
    <input type="text" />
    <button value="save" />
    <button value="delete" />
  </form>
)

test('Parses the props', () => {
  const component = mount(
    <Permit
      permissions={640}
      editable={<Editable />}
      viewable={<Viewable />}
      executable={<Executable />}
      role="admin"
      roles={{
        admin: 0,
        editor: 1,
        guest: 2,
      }}
    />,
  )

  expect(component.instance().getPermissions()).toEqual([
    (6).toString(2).padStart('3', '0'),
    (4).toString(2).padStart('3', '0'),
    (0).toString(2).padStart('3', '0'),
  ])
  expect(component.instance().getRole()).toEqual(0)
  expect(component.instance().getAccess()).toEqual({ r: 1, w: 1, x: 0 })

  component.setProps({ role: 'editor' })
  component.setProps({ permissions: 740 })

  expect(component.instance().getPermissions()).toEqual([
    (7).toString(2).padStart('3', '0'),
    (4).toString(2).padStart('3', '0'),
    (0).toString(2).padStart('3', '0'),
  ])
  expect(component.instance().getRole()).toEqual(1)
  expect(component.instance().getAccess()).toEqual({ r: 1, w: 0, x: 0 })
})

test('Renders the components passed in props', () => {
  const component = mount(
    <Permit
      permissions={744}
      editable={<Editable />}
      viewable={<Viewable />}
      executable={<Executable />}
      role="admin"
      roles={{
        admin: 0,
        editor: 1,
        guest: 2,
      }}
    />,
  )

  expect(component.props().permissions).toEqual(744)
  expect(toJSON(component)).toMatchSnapshot()

  component.setProps({ role: 'editor' })
  expect(component.props().role).toEqual('editor')
  expect(toJSON(component)).toMatchSnapshot()

  component.setProps({ role: 'guest' })
  expect(component.props().role).toEqual('guest')
  expect(toJSON(component)).toMatchSnapshot()
})

test('Renders the children', () => {
  const component = mount(
    <Permit
      permissions={600}
      role="admin"
      roles={{
        admin: 0,
        editor: 1,
        guest: 2,
      }}
    >
      <span>Hello Admin!</span>
    </Permit>,
  )

  expect(component.props().permissions).toEqual(600)
  expect(toJSON(component)).toMatchSnapshot()

  component.setProps({ role: 'editor' })
  expect(component.props().role).toEqual('editor')
  expect(toJSON(component)).toMatchSnapshot()

  component.setProps({ role: 'guest' })
  expect(component.props().role).toEqual('guest')
  expect(toJSON(component)).toMatchSnapshot()
})
