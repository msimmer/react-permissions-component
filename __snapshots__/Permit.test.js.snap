// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the children 1`] = `
<Permit
  permissions={600}
  role="admin"
  roles={
    Object {
      "admin": 0,
      "editor": 1,
      "guest": 2,
    }
  }
>
  <span>
    Hello Admin!
  </span>
</Permit>
`;

exports[`Renders the children 2`] = `
<Permit
  permissions={600}
  role="editor"
  roles={
    Object {
      "admin": 0,
      "editor": 1,
      "guest": 2,
    }
  }
/>
`;

exports[`Renders the children 3`] = `
<Permit
  permissions={600}
  role="guest"
  roles={
    Object {
      "admin": 0,
      "editor": 1,
      "guest": 2,
    }
  }
/>
`;

exports[`Renders the components passed in props 1`] = `
<Permit
  editable={<Editable />}
  executable={<Executable />}
  permissions={744}
  role="admin"
  roles={
    Object {
      "admin": 0,
      "editor": 1,
      "guest": 2,
    }
  }
  viewable={<Viewable />}
>
  <Executable>
    <form>
      <input
        type="text"
      />
      <button
        value="save"
      />
      <button
        value="delete"
      />
    </form>
  </Executable>
</Permit>
`;

exports[`Renders the components passed in props 2`] = `
<Permit
  editable={<Editable />}
  executable={<Executable />}
  permissions={744}
  role="editor"
  roles={
    Object {
      "admin": 0,
      "editor": 1,
      "guest": 2,
    }
  }
  viewable={<Viewable />}
>
  <Viewable>
    <form>
      <input
        readOnly={true}
        type="text"
      />
    </form>
  </Viewable>
</Permit>
`;

exports[`Renders the components passed in props 3`] = `
<Permit
  editable={<Editable />}
  executable={<Executable />}
  permissions={744}
  role="guest"
  roles={
    Object {
      "admin": 0,
      "editor": 1,
      "guest": 2,
    }
  }
  viewable={<Viewable />}
>
  <Viewable>
    <form>
      <input
        readOnly={true}
        type="text"
      />
    </form>
  </Viewable>
</Permit>
`;
